### DISCUSSION TITLE
Task 2 function

### DISCUSSION THREAD

**Discussion by kg492** [Part-II-Computational-Physics/.github#40] [General]
I tried defining the CompleteHamiltonian2 as:

def CompleteHamiltonian2(N, h=1.):
    
    Hloc =  - ft.reduce(np.kron, [sigmaZ,sigmaZ]) - 0.5*h*ft.reduce(np.kron, [sigmaX,I]) - 0.5*h*ft.reduce(np.kron, [I,sigmaX])
    H = np.zeros((2**N, 2**N))
    for ii in range(N):
        H -= ft.reduce(np.kron, Hloc)
        H -= np.roll(Hloc, 1, axis=0)

    return H
CompleteHamiltonian2(2)

as I think just a sum of Hloc terms? Not really sure how to get it going further, what to think about to get the correct expression, I checked previous discussions but feel like some ppl took very different approach.  Any help?

**Comment by JNHallen** [comment]
Hi! Let's see if I can give you some pointers. 

Firstly, I suggest you define Hloc as a $4 \times 4$ matrix, matching the example given. 

The trick to use in this question is that the Kronecker product of $K$ $2\times 2$ identity matrices is simply a larger identity matrix, of size $2^K \times 2^K$. We want to write our Hamiltonian as a sum of local terms as you say, and the $i^\mathrm{th}$ term in that sum is 

$$I_1 ... I_{i-1} {\cal H}^{\rm loc}_{i,i+1} I_{i+2} ... I_N$$

which, in our code we could then write as 
`np.kron(np.identity(2**(i-1)), np.kron(sigmaZ, np.identity(2**(N- i-1))))`

You can add up terms like that in your function to construct the Hamiltonian.
